src/EventStore.Core/DataStructures/ProbabilisticFilter/PersistentBloomFilters/BloomFilterAccessor.cs

```diff
@@ -0,0 +1,226 @@
+			FileSize += cacheLineSize;
+
+			// header is not part of the pages
+			// last page may be a partial page
+			var dataSize = FileSize - CacheLineSize;
+			NumPages = dataSize / PageSize;
+			if (dataSize % PageSize != 0)
+				NumPages++;
+		}
+
+		public long LogicalFilterSize { get; }
+		public long LogicalFilterSizeBits { get; }
+		public int CacheLineSize { get; }
+		public int HashSize { get; }
+		public int PageSize { get; }
+		public long FileSize { get; }
+
+		// last page may be a partial page
+		public long NumPages { get; }
+
+		// FileSize is definitely a multiple of CacheLineSize
+		public long NumCacheLines => FileSize / CacheLineSize;
+
+		// Pointer to the beginning of the file
+		private byte* _pointer;
+		public byte* Pointer {
+			get => _pointer;
+			set {
+				if ((long)value % CacheLineSize != 0) {
+					//qq consider whether to keep this.. throw? log error?
+					// do at least test out how the error message comes out
+					_log.Error("Pointer {pointer} was not aligned to a cacheline {cacheLineSize}", (long)value, CacheLineSize);
+					throw new InvalidOperationException($"Pointer {(long)value} was not aligned to a cacheline {CacheLineSize}");
+				}
+				_pointer = value;
+			}
+		}
+
+		// converts a bit position in the logical filter into a byte position in the file
+		public long GetBytePositionInFile(long bitPosition) {
+			var bytePositionInFile = bitPosition / 8;
+
+			// account for 4 bytes at the end of each cache line being a hash
+			bytePositionInFile += bytePositionInFile / (CacheLineSize - HashSize) * HashSize;
+
+			// first cache line is reserved for the header
+			bytePositionInFile += CacheLineSize;
+
+			return bytePositionInFile;
+		}
+
+		// converts page number into byte position of the page in the file
+		public (long PositionInFile, int PageSize) GetPagePositionInFile(int pageNumber) {
+			// first cache line is reserved for the header
+			var positionInFile = CacheLineSize + (pageNumber * PageSize);
+
+			var pageSize = Math.Min(PageSize, (int)(FileSize - positionInFile));
+
+			return (positionInFile, pageSize);
+		}
+
+		// determines which page a bytePosition in the file is for.
+		public long GetPageNumber(long bytePositionInFile) {
+			// pages start after the header
+			return (bytePositionInFile - CacheLineSize) / PageSize;
+		}
+
+		public void EnsureBounds(long bytePosition, int length) {
+			if (bytePosition > FileSize - length) {
+				throw new ArgumentOutOfRangeException(
+					nameof(bytePosition),
+					bytePosition,
+					$"Position {bytePosition:N0} + length {length:N0} = {bytePosition + length:N0} out of file size {FileSize:N0}");
+			}
+		}
+
+		public ref byte ReadByte(long bytePositionInFile) {
+			EnsureBounds(bytePositionInFile, length: 1);
+			return ref *(Pointer + bytePositionInFile);
+		}
+
+		public Span<byte> ReadCacheLineFor(long bytePositionInFile) {
+			bytePositionInFile = bytePositionInFile / CacheLineSize * CacheLineSize;
```

Comment:
suggestion: use .RoundDownToMultipleOf(CacheLineSize); for consistency

Reply:
👍